Convert formatting from force

njgheorghita opened the pull request:
### What was wrong?
`.formatting` had functions that depended on deprecated `force_*` functions

### How was it fixed?
update `.formatting` to use `to_*` functions

#### Cute Animal Picture
![image](https://user-images.githubusercontent.com/9753150/36003185-2249b166-0cea-11e8-99ac-971a5b673763.png)

njgheorghita commented:
@pipermerriam @carver not sure if reorganization is in order here? went with inlining the dependencies inside `.conversions`, but if there's a better/more preferred way to deal with the circular dependencies b/w `.conversions` & `.hexidecimal` i'll make the fix

carver commented:
Oh, and the fact that the tests pass means that we're missing some...

njgheorghita commented on eth_utils/formatting.py:
```diff
@@ -13,7 +13,9 @@ def pad_left(value, to_size, pad_with):
-    pad_with_value = force_bytes(pad_with) if is_bytes(value) else force_text(pad_with)
+    pad_args = {'primitive': pad_with} if is_bytes(pad_with) else {'text': pad_with}
+    pad_value = to_bytes(**pad_args) if is_bytes(value) else to_text(**pad_args)
+    pad_with_value = to_text(text=pad_value)
```
@carver I was unsure about this too - but the tests pass and it seems like all the necessary edge cases are represented. 
```python
@pytest.mark.parametrize(
    'value,to_size,pad_with,expected',
    (
        ('', 10, '0', '0000000000'),
        ('', 10, b'0', '0000000000'),
        (b'', 10, '0', b'0000000000'),
        (b'', 10, b'0', b'0000000000'),
        ('1', 10, '0', '0000000001'),
        ('1', 5, '123', '12311'),
        ('12345', 4, '0', '12345'),
    )
)

from what I can tell - `.rjust` & `.ljust` only work when `pad_with` is a single character. there are some test cases where `pad_with` is multiple characters. so we can definitely make that change depending on whether or not we want to support multi-char `pad_with` values

carver replied:
How was `(b'', 10, '0', b'0000000000'),` case working on the code when `pad_with_value` was being forced to text?

carver replied:
I know this question seems moot now, but I'd like to know, because it implies that we aren't testing the code path that we thought we were.

njgheorghita replied:
that was my bad - pad_value was getting passed to `to_text()` as a text arg which implicitly believes it's a text value and just returns it - so a bytes value getting passed in as a text arg would simply be returned as that bytes value

carver replied:
> that was my bad - pad_value was getting passed to to_text() as a text arg which implicitly believes it's a text value and just returns it

That's really interesting. I would characterize that as a bug in `to_text()`. It should validate and fail if the `text` parameter is not a `str`.

njgheorghita commented on eth_utils/formatting.py:
```diff
@@ -26,14 +28,16 @@ def pad_right(value, to_size, pad_with):
-    )
+    prefix_args = {'primitive': prefix} if is_bytes(prefix) else {'text': prefix}
+    prefix_value = to_bytes(**prefix_args) if is_bytes(value) else to_text(**prefix_args)
+    return value.startswith(prefix_value)
```
56 test cases that break on this change - it seems like some other util functions don't make this assumption : `.hexidecimal` - `is_hex()` and `is_prefixed()`, so i think the tradeoff here would be updating those functions to accommodate this assumption

carver replied:
I seem to remember another thread where we decided that `is_hex` should only accept string arguments, for example: that `b'0x12'` should *not* be treated as hex. So I think updating those other functions is the right thing.

pipermerriam replied:
+1 to @carver 's suggestions.  I think these changes are appropriate.

- require `value` and the padding to be of the same type.
- establish that hex values **must** be text types.

njgheorghita replied:
cool - so would establishing all hex values must be text types apply to the entire repo? for example in `.hexidecimal` would something like this work? 
```python3
def is_0x_prefixed(value):
    if is_bytes(value):
        return is_prefixed(value, b'0x') or is_prefixed(value, b'0X')
    return is_prefixed(value, '0x') or is_prefixed(value, '0X')
```
which works fine, or should it simply be 
```python3
def is_0x_prefixed(value):
    return is_prefixed(value, '0x') or is_prefixed(value, '0X')
```
which causes a fair amount of breaking tests - but if i'm understanding correctly maintains the assumption that all possible hex values would be text type

njgheorghita replied:
@pipermerriam also at this point `.formatting` is looking pretty scarce
```python
def pad_left(value, to_size, pad_with):
    """
    Should be called to pad value to expected length
    """
    return value.rjust(to_size, pad_with)

def pad_right(value, to_size, pad_with):
    """
    Should be called to pad value to expected length
    """
    return value.ljust(to_size, pad_with)

def is_prefixed(value, prefix):
    return value.startswith(prefix)
```
@carver suggested deprecating these functions and using the stdlib functions - which i think makes sense now

pipermerriam replied:
Re `is_0x_prefixed`.  Lets only support text typed values and explicitely raise if the value is not a text type.

```
def is_0x_prefixed(value):
    if not is_text(value):
        raise TypeError("is_0x_prefixed requires text typed arguments.  Got: {0}".format(type(value)))
    return is_prefixed(value, '0x') or is_prefixed(value, '0X')
```

pipermerriam replied:
> @carver suggested deprecating these functions and using the stdlib functions - which i think makes sense now

+1

njgheorghita replied:
@pipermerriam  in terms of deprecating the `.formatting` utils - do you want to just remove them entirely (like the plan is with `.string` utils once we convert everything over to using `.conversion` utils) or should i create a deprecation warning and attach it to each function

carver replied:
At first glance: I think the deprecation goes in v0, and they are removed from v1.

njgheorghita commented:
@carver @pipermerriam thanks for your help! I got rid of the deprecated `.formatting` utils in this pr. Added a deprecation warning to them in #56 .  
updated `is_0x_prefixed()` & `is_hex()` to only accept text types. 
will continue removing the `force_*` decorators from `.hexidecimal` in the next pr and that should wrap it up for the switch from `force_*` -> `to_*`.
if there's something else i missed in this pr, holler

carver commented on eth_utils/hexidecimal.py:
```diff
@@ -31,11 +30,15 @@ def decode_hex(value):
-    return is_prefixed(value, '0x') or is_prefixed(value, '0X')
+    if not is_text(value):
+        raise TypeError(
+            "is_0x_prefixed requires text typed arguments. Got: {0}".format(type(value))
```
My favorite way to show the value that triggered the error is a `repr(value)`, so that the caller can immediately see both the value and type.

Just sharing, it's up to you.

njgheorghita replied:
makes sense to me, thanks

carver commented on tests/hexidecimal-utils/test_0x_prefix_addition_and_removal.py:
```diff
@@ -27,8 +25,6 @@ def test_add_0x_prefix(value, expected):
     (
         ('', ''),
-        (b'', b''),
-        (b'0x', b''),
```
Test that raises

carver commented on tests/hexidecimal-utils/test_0x_prefix_addition_and_removal.py:
```diff
@@ -10,8 +10,6 @@
     (
         ('', '0x'),
-        (b'', b'0x'),
-        (b'0x', b'0x'),
```
We should add a test that `bytes` input raises the `TypeError`

njgheorghita commented on eth_utils/hexidecimal.py:
```diff
@@ -31,11 +30,15 @@ def decode_hex(value):
     if not is_string(value):
         raise TypeError('Value must be an instance of str or unicode')
-    return add_0x_prefix(codecs.encode(value, 'hex'))
+    return add_0x_prefix(to_text(codecs.encode(value, 'hex')))
```
haha - was thoroughly convinced `hexidecimal` was the proper spelling, but yeah i'll make the change next

carver replied:
Hah, yeah, I occasionally mix it up, too. My mnemonic is that it's spelled the same way as "hexagon".

pipermerriam commented:
Looks good. Merge and release next v1 beta with this when ready.

carver commented:
I was about to do a release, then realized that I'd like to see #67 merged before the next release. Without that change, `to_hex(value=b'\xff')` raises a TypeError.
